core/frontend.ml

```diff
@@ -25,89 +25,178 @@ struct
 
   let session_exceptions = Settings.get_value Basicsettings.Sessions.exceptions_enabled
 
-  (* (These functions correspond to 'first' in an arrow) *)
-  let after_typing f (a, b, c) = (f a, b, c)
-  let _after_alias_expansion f (a, b) = (f a, b)
-  let extension enabled f g (a, b, c) =
-    if enabled
-    then f g (a, b, c)
-    else (a, b, c)
-
-  let after_typing_ext enabled = extension enabled after_typing
-  let before_typing_ext enabled f x = if enabled then f x else x
-
-  let program_pipeline =
-    fun tyenv pos_context program ->
-      let program = (ResolvePositions.resolve_positions pos_context)#program program in
-      let _program = CheckXmlQuasiquotes.checker#program program in
-      let () = DesugarSessionExceptions.settings_check program in
-      let apply =
-        ( DesugarModules.desugar_program
-          ->- before_typing_ext session_exceptions DesugarSessionExceptions.wrap_linear_handlers
-          ->- DesugarLAttributes.desugar_lattributes#program
-          ->- LiftRecursive.lift_funs#program
-          ->- DesugarDatatypes.program tyenv
-          ->- uncurry TypeSugar.Check.program
-          (*->- after_typing ((FixTypeAbstractions.fix_type_abstractions tyenv)#program ->- snd3)*)
-          ->- after_typing ((DesugarCP.desugar_cp tyenv)#program ->- snd3)
-          ->- after_typing ((DesugarInners.desugar_inners tyenv)#program ->- snd3)
-          ->- after_typing_ext session_exceptions ((DesugarSessionExceptions.insert_toplevel_handlers tyenv)#program ->- snd3)
-          ->- after_typing_ext session_exceptions ((DesugarSessionExceptions.desugar_session_exceptions tyenv)#program ->- snd3)
-          ->- after_typing ((DesugarProcesses.desugar_processes tyenv)#program ->- snd3)
-          ->- after_typing ((DesugarFors.desugar_fors tyenv)#program ->- snd3)
-          ->- after_typing ((DesugarRegexes.desugar_regexes tyenv)#program ->- snd3)
-          ->- after_typing ((DesugarFormlets.desugar_formlets tyenv)#program ->- snd3)
-          ->- after_typing ((DesugarPages.desugar_pages tyenv)#program ->- snd3)
-          ->- after_typing ((DesugarFuns.desugar_funs tyenv)#program ->- snd3))
-      in
-      let (program, _typ, _tenv) as result = apply program in
-      (result, ModuleUtils.get_ffi_files program)
-
-
-let program tyenv pos_context program =
+  let repeat_type_check =
+    Settings.get_value
+      Basicsettings.TypeSugar.check_frontend_transformations
+
+
+  let for_side_effects ignored_transformer program  =
+    let _ = ignored_transformer program in
+    program
+
+  type pre_typing_program_transformer =
```

Comment:
Can you organise these into logical units (coherent modules)? It seems to me, that we need an abstraction for "transformers".